Add support for Docker

afirth opened the pull request:
1) ./Dockerfile, which builds the server with mod_wsgi-express. Data and config.py are not placed.
2) demo/Dockerfile, which adds the example data and config.py to 1)
3) demo_light/Dockerfile, which builds only the python, as the demo in
existing docs

afirth commented:
Again referencing #305, and in the spirit of #344, this adds support for deployment through mod_wsgi-express, leveraging upstream code as much as possible. The mod_wsgi author publishes mod_wsgi-docker at https://github.com/GrahamDumpleton/mod_wsgi-docker, which this is built on top of. I would welcome further input or pulls. Images are available on dockerhub:
1) afirth/ga4gh_debian_apache
2) afirth/ga4gh_debian_apache_demo
3) afirth/ga4gh_centos_demo

I will add docs shortly. @hershman, what do you think? I didn't see your PR until I had these mostly done last night.

jeromekelleher commented:
This looks really good @afirth, thanks for putting this together. As you've noticed, it's unfortunate that @hershman has already opened a PR with a working Docker recipe. I think the fairest thing to do is to merge @hershman's changes first, and perhaps you could extend on this afterwards? I like the idea of using the upstream images, this is a good idea.

afirth commented:
I'd love to hear from @hershman, I'd love to collaborate rather than squashing each other's work. I'm sure there are some design considerations he's thought of that I haven't. I'll drop a comment on #344.

afirth commented:
Added documentation. Now supports using environment variables to set data_source and debug. Automated builds available at https://registry.hub.docker.com/u/afirth/ga4gh_server_apache/
Updated documentation. Demo and production behind apache can now be run with one command. Support use of host-side dataset. Pull in @hershman's Dockerfile as well

pgrosu commented:
Neat automation and integration with documentation - thanks for the nice work.

~p

dcolligan commented on docs/source/installation.rst:
```diff
@@ -157,5 +157,102 @@ error log (in Apache on Debian/Ubuntu, for example, this is ``/var/log/apache2/e
+
+  $ VBoxManage controlvm boot2docker-vm natpf1 "ga4gh,tcp,127.0.0.1,8000,,8000"
+
+For more info on port forwarding see https://www.virtualbox.org/manual/ch06.html#natforward and https://github.com/CenturyLinkLabs/panamax-ui/wiki/How-To%3A-Port-Forwarding-on-VirtualBox
```
these links should use the sphinx link syntax

afirth commented:
@dcolligan I'm not sure what you mean. I didn't think link text added anything here so I left them plain. No markup is required for links without alt-text. Is there a guideline that all links should be hidden?

dcolligan commented:
@afirth  It breaks up the flow of reading to see long gnarly URLs in the midst of a text block.  "... see here and here." would be easier for a reader to parse, for instance.

afirth commented:
@dcolligan updated, thanks for the review

dcolligan commented:
To give a proper review, I think we should have someone that actually knows something about docker review this.  But if everyone else is just as inexperienced with it as I am, I suppose I can look at it a bit more closely after we resolve which parts of which of the 2 PRs we want to merge.

afirth commented:
As I said, I've merged #344 into here as there's been no movement on that build, and it was/is failing. I'd like to get this in so I can switch the autobuilds over to upstream instead of my branch

afirth commented:
Oops, looks like it was rebased a few minutes ago. I can rebase on top of it and fix the doc merge conflicts if that's the way others want to go.

hershman commented:
I @afirth - this is really great.  I'm not expert enough on Docker to comment on all your changes.  Given that #344 has been reviewed and that the discussion here may continue, I'd prefer to get that pulled in and have this squash it when ready.

jeromekelleher commented:
OK, we've merged @hershman's PR, and you should be able to build on that now @afirth. Can you rebase this PR off upstream/develop please? We can review the changes (to the degree that we're able to --- I agree with @dcolligan's point above, I'm also clueless in terms of Docker) then.

afirth commented:
rebased and squashed

jeromekelleher commented on .whiskey/action_hooks/pre-build:
```diff
@@ -0,0 +1,7 @@
+#!/usr/bin/env bash
```
What does this file do?

afirth replied:
It is a pre-build hook which adds the source of this repository to requirements.txt. Basically as doing ``pip install -e ./`` Previous build instructions pull from github rather than building from source. If the user chooses the automated builds, that will still happen, but this way it is possible to build from source locally for development. I will expand the comments here to indicate the file's purpose.

jeromekelleher commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+############################################################
```
I think we should keep all the Docker specific stuff in the `deploy` directory.

afirth replied:
By convention the dockerfile is in the base of the project. In this case it is also technically required because the mod_wsgi-dockerfile copies the CWD to the docker container. Res: won't fix.

jeromekelleher commented on deploy/config.py:
```diff
@@ -0,0 +1,11 @@
+#TODO this logic could move to frontend.configure() or BaseConfig
+# For docker, if/when a default is set in serverconfig.py, run -v /localdata:/default-path
+# If the env variable GA4GH_DATA_SOURCE is set, use that path. Otherwise, use the default path
+DATA_SOURCE = os.getenv('GA4GH_DATA_SOURCE', "/ga4gh-example-data")
```
I think we support configuration from environment variables already --- I don't quite get what we're trying to do here.

afirth replied:
The only environment variable imported is GA4GH_CONFIGURATION, the path of the config file. I agree we should support this in frontend, but I thought that would be out of scope of adding support for Docker. This allows setting the location of the datasource by passing an env variable to the docker container. os.getenv returns arg2 if arg1 is unset. This logic could be discarded when frontend.py is updated with production oriented argument parsing. Res: won't fix

jeromekelleher commented on docs/source/installation.rst:
```diff
@@ -157,26 +157,102 @@ error log (in Apache on Debian/Ubuntu, for example, this is ``/var/log/apache2/e
 .. code-block:: bash
 
-  $ docker build .
+  $ docker run -e GA4GH_DATA_SOURCE=/data -v /my/ga4gh_data/:/data:ro -itd -p 8000:80 --name ga4gh_server afirth/ga4gh_apache_server:prod
```
Can you explain what's happening here please? I'm sorry, I'm a complete docker noob, so I'm struggling to understand how  the config files and the environment variables are fitting together.

afirth replied:
Absolutely. Res: fixed

jeromekelleher commented on docs/source/installation.rst:
```diff
@@ -157,26 +157,102 @@ error log (in Apache on Debian/Ubuntu, for example, this is ``/var/log/apache2/e
 
-First build the Docker Image.  From the ``deploy`` folder run:
+First, you need an environment running the docker daemon. For non-production use, we recommend `boot2docker <http://boot2docker.io/>`_. For production use you should install docker on a stable linux distro.
+`Platform specific Docker installation instructions <https://docs.docker.com/installation/) (OS X and Windows are instructions for boot2docker>`_
```
I think this link/sentence is broken.

afirth replied:
Badly broken. Thanks. Res:fixed

jeromekelleher commented on docs/source/installation.rst:
```diff
@@ -157,26 +157,102 @@ error log (in Apache on Debian/Ubuntu, for example, this is ``/var/log/apache2/e
+
+This will run the docker container in the background, and translate calls from your host environment
+port 8000 to the docker container port 80. At that point you should be able to access it like a normal website, albeit on port 8000.
+Running in boot2docker, you will need to forward the port from the boot2docker vm to the host.
```
It would be nice to link to boot2docker here, in case the reader wasn't paying attention earlier.

Also, s/vm/VM

afirth replied:
Agreed. Res: fixed.

jeromekelleher commented on docs/source/installation.rst:
```diff
@@ -157,26 +157,102 @@ error log (in Apache on Debian/Ubuntu, for example, this is ``/var/log/apache2/e
 
-This will run the docker container in the background, and translate calls from your host environment
-port 8888 to the docker container port 80. At that point you should be able to access it like a normal website.
+For more info on port forwarding see `the VirtualBox manual <https://www.virtualbox.org/manual/ch06.html#natforward>`_ and this `wiki article <https://github.com/CenturyLinkLabs/panamax-ui/wiki/How-To%3A-Port-Forwarding-on-VirtualBox>`_.
```
This is all a bit Mac specific --- you're assuming the user is running virtualbox on a mac? I'm happy to have Mac specific instructions, but I think they should be clearly signposted as such. I'm not sure we need to include port-forwarding instructions on this page really, since it says at the top that the audience is system administrators who are looking to deploy the server in production.

afirth replied:
Not assuming, but this step will be omitted if running directly on a linux host. If running on mac or windows, boot2docker is the supported installation option for docker (https://docs.docker.com/installation/mac/), and boot2docker *includes* virtualbox. It's just a copy of virtualbox and a tiny linux vm with docker installed.

I agree with you that the instructions differ a bit from the stated purpose of the page, but it no longer makes sense to me to split demo + production as the differences are so minor with the docker build. I have tried to make these instructions accessible to researches and contributors who are not sysadmins by including the roadblocks I ran into installing on three different configurations, and I'd hope if others find more roadblocks they can add them here too.

jeromekelleher commented:
I've had a look through this @afirth, and it's generally great. I don't quite understand how it all fits together though (which is my fault as a Docker noob), so it would be really helpful if you could write a quick description of how it all works, perhaps in a README in the deploy directory.

The documentation looks good, but I wonder if we're mixing the message of the page a little bit with all the details of running demos, fixing DNS problems and so on. This page is intended to be the definitive source of information for production deployments, but there's an awful lot of stuff here for getting things working on your laptop. Perhaps this sort of information would also be better in the Docker README?

Anyway, this all excellent, so thanks a lot!

afirth commented:
@jeromekelleher Thanks a lot for your comments. I've fixed those that I can and will squash the commits after review.

Regarding the README, I worry that this is akin to explaining how apache and mod_wsgi work in this repo. It would duplicate the documentation at https://github.com/GrahamDumpleton/mod_wsgi-docker which is the base of this development (from the mod_wsgi author), and on docs.docker.com. I've added an explicit link to the upstream project to the advanced section of the docs.

jeromekelleher commented:
Thanks very much for the responses @afirth, and for your patience with stupid questions! I like this a lot, I think we should merge and see how it goes as we get more experience. We can streamline things a bit later if needs be.

+1 after squashing.

afirth commented:
squashed

jeromekelleher commented on ga4gh/application.wsgi:
```diff
@@ -0,0 +1,3 @@
+from ga4gh.frontend import app as application
```
Sorry, I just spotted this now as I was about to merge it. I don't think this file should be in the ga4gh directory, as it's not part of the Python package. Could you move this to deploy (or somewhere else) please?

afirth replied:
Absolutely. I'll leave that question for another day :) 
Moved the file and copied it in place with the pre-build hook. Diff [here](https://github.com/UMMS-Biocore/server/commit/5785743e401f028cbf323fe0186bcc04d031a41d), squashed into this PR

afirth commented:
@jeromekelleher No problem. Squashed on top, dockerhub builds passed.

jeromekelleher commented:
Merging. Thanks @afirth!
